File: backend/src/dearth/strategies/ModrinthFilter.ts
import { IFilterStrategy, IFileInfo } from "../types.js";
import { logger } from "../../utils/logger.js";

interface IModrinthProject {
  client_side: string;
  server_side: string;
  project_type: string;
  categories: string[];
}

export class ModrinthFilter implements IFilterStrategy {
  name = "ModrinthFilter";
  private readonly API_BASE = "https://api.modrinth.com/v2";

  private extractProjectId(infos: { name: string; data: string }[]): string | null {
    for (const info of infos) {
      if (info.name === "modrinth.index.json" || info.name === "modrinth.json") {
        try {
          const data = JSON.parse(info.data);
          return data.project_id || null;
        } catch {
          continue;
        }
      }
    }
    return null;
  }

  private async fetchProjectInfo(projectIds: string[]): Promise<Map<string, IModrinthProject>> {
    const projectMap = new Map<string, IModrinthProject>();
    const batchSize = 50;

    for (let i = 0; i < projectIds.length; i += batchSize) {
      const batch = projectIds.slice(i, i + batchSize);
      const idsParam = batch.join(",");

      try {
        const response = await fetch(`${this.API_BASE}/projects?ids=${encodeURIComponent(idsParam)}`, {
          headers: {
            'User-Agent': 'DeEarthX-V3/1.0.0'
          }
        });

        if (response.ok) {
          const projects: Array<any> = await response.json();

          for (const project of projects) {
            if (project && project.id) {
              projectMap.set(project.id, {
                client_side: project.client_side,
                server_side: project.server_side,
                project_type: project.project_type,
                categories: project.categories || []
              });
            }
          }
        }
      } catch (error) {
        logger.error("获取 Modrinth 项目信息失败", { error, batchSize });
      }
    }

    return projectMap;
  }

  private isClientMod(project: IModrinthProject): boolean {
    const clientSide = project.client_side;
    const serverSide = project.server_side;

    return (
      clientSide === "required" ||
      (clientSide === "optional" && serverSide === "unsupported")
    );
  }

  async filter(files: IFileInfo[]): Promise<string[]> {
    const clientMods: string[] = [];
    const projectIds: Array<{ filename: string; projectId: string }> = [];

    for (const file of files) {
      const projectId = this.extractProjectId(file.infos);
      if (projectId) {
        projectIds.push({ filename: file.filename, projectId });
      }
    }

    if (projectIds.length === 0) {
      logger.info("未找到 Modrinth 项目 ID");
      return clientMods;
    }

    logger.info(`找到 ${projectIds.length} 个 Modrinth 项目`, { 数量: projectIds.length });

    const uniqueProjectIds = [...new Set(projectIds.map(p => p.projectId))];
    const projectMap = await this.fetchProjectInfo(uniqueProjectIds);

    for (const { filename, projectId } of projectIds) {
      const project = projectMap.get(projectId);
      if (project && this.isClientMod(project)) {
        clientMods.push(filename);
      }
    }

    logger.info("Modrinth 筛选完成", { 客户端模组数: clientMods.length });
    return clientMods;
  }
}

File: backend-net/src/DeEarthX.Dearth/ModrinthFilter.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using DeEarthX.Core.Abstractions;
using DeEarthX.Infrastructure.Http;

namespace DeEarthX.Dearth;

public sealed class ModrinthFilter : IBatchFilterStrategy
{
    private const string ApiBase = "https://api.modrinth.com/v2";
    private const int BatchSize = 100;

    private readonly IDeEarthXHttpService _http;
    private readonly ILogService _log;
    private readonly Dictionary<string, ModrinthProjectInfo> _cache = new(StringComparer.Ordinal);

    public ModrinthFilter(IDeEarthXHttpService http, ILogService log)
    {
        _http = http;
        _log = log;
    }

    public string Name => "modrinth";

    public async Task<HashSet<string>> FilterBatchAsync(List<ModFileInfo> files)
    {
        var result = new HashSet<string>(StringComparer.Ordinal);
        if (files.Count == 0)
        {
            return result;
        }

        var pairs = new List<(string FilePath, string ProjectId)>(files.Count);
        foreach (var file in files)
        {
            var pid = ExtractProjectId(file.Infos);
            if (!string.IsNullOrEmpty(pid))
            {
                pairs.Add((file.FilePath, pid));
            }
        }

        if (pairs.Count == 0)
        {
            return result;
        }

        var uniqueIds = pairs.Select(p => p.ProjectId).Distinct(StringComparer.Ordinal).ToList();
        var projectMap = await FetchProjectInfoAsync(uniqueIds).ConfigureAwait(false);

        foreach (var (filePath, pid) in pairs)
        {
            if (projectMap.TryGetValue(pid, out var info) && IsClientMod(info))
            {
                result.Add(filePath);
            }
        }

        return result;
    }

    private static string? ExtractProjectId(List<InfoFile> infos)
    {
        foreach (var info in infos)
        {
            if (!info.Name.Equals("modrinth.index.json", StringComparison.OrdinalIgnoreCase) &&
                !info.Name.Equals("modrinth.json", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            try
            {
                using var doc = JsonDocument.Parse(info.Data);
                if (doc.RootElement.TryGetProperty("project_id", out var pidEl) &&
                    pidEl.ValueKind == JsonValueKind.String)
                {
                    return pidEl.GetString();
                }
            }
            catch
            {
                continue;
            }
        }

        return null;
    }

    private async Task<Dictionary<string, ModrinthProjectInfo>> FetchProjectInfoAsync(List<string> projectIds)
    {
        var map = new Dictionary<string, ModrinthProjectInfo>(StringComparer.Ordinal);
        var uncached = new List<string>();
        foreach (var id in projectIds)
        {
            if (_cache.TryGetValue(id, out var cached))
            {
                map[id] = cached;
            }
            else
            {
                uncached.Add(id);
            }
        }

        if (uncached.Count == 0)
        {
            return map;
        }

        for (var i = 0; i < uncached.Count; i += BatchSize)
        {
            var batch = uncached.Skip(i).Take(BatchSize).ToList();
            var idsParam = Uri.EscapeDataString(string.Join(",", batch));
            try
            {
                var projects = await _http.GetJsonAsync<List<ModrinthProjectInfo>>(
                    $"{ApiBase}/projects?ids={idsParam}").ConfigureAwait(false);

                if (projects is null)
                {
                    continue;
                }

                foreach (var p in projects)
                {
                    if (!string.IsNullOrEmpty(p.Id))
                    {
                        map[p.Id] = p;
                        _cache[p.Id] = p;
                    }
                }
            }
            catch (Exception ex)
            {
                _log.Error("获取 Modrinth 项目信息失败", ex);
            }
        }

        return map;
    }

    private static bool IsClientMod(ModrinthProjectInfo project)
    {
        return project.ClientSide == "required" ||
               (project.ClientSide == "optional" && project.ServerSide == "unsupported");
    }

    public sealed class ModrinthProjectInfo
    {
        public string Id { get; set; } = string.Empty;

        [JsonPropertyName("client_side")]
        public string ClientSide { get; set; } = string.Empty;

        [JsonPropertyName("server_side")]
        public string ServerSide { get; set; } = string.Empty;

        [JsonPropertyName("project_type")]
        public string? ProjectType { get; set; }

        public List<string>? Categories { get; set; }
    }
}
